core/src/lib/thread_pool.h

```diff
@@ -0,0 +1,143 @@
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+   02110-1301, USA.
+*/
+/**
+ * @file
+ * simple thread pool
+ */
+
+#ifndef BAREOS_LIB_THREAD_POOL_H_
+#define BAREOS_LIB_THREAD_POOL_H_
+
+#include <vector>
+#include <thread>
+#include <optional>
+#include <deque>
+#include <mutex>
+#include <shared_mutex>
+#include <condition_variable>
+#include <future>
+#include <type_traits>
+#include <exception>
+
+#include "lib/thread_util.h"
+
+struct worker_pool {
+  std::vector<std::thread> threads{};
+  std::size_t num_borrowed{0};
+  std::size_t min_workers{0};
+  std::size_t dead_workers{0};
+
+  std::size_t num_actual_workers() const
+  {
+    return threads.size() - num_borrowed - dead_workers;
+  }
+};
+
+class thread_pool {
+ public:
+  // fixme(ssura): change this to std::move_only_function (c++23)
+  using task = std::function<void()>;
+
+  thread_pool(std::size_t num_threads = 0);
+
+  void ensure_num_workers(std::size_t num_threads);
+  void finish();   // wait until all submitted tasks are accomplished
+  ~thread_pool();  // stops as fast as possible; dropping outstanding tasks
+
+  void enqueue(task&& t);  // queue task to be worked on by worker threads;
+                           // should not block
+  void borrow_thread(task&& t);  // steal thread to work on task t; can block
+
+ private:
+  std::condition_variable worker_death;
+  synchronized<worker_pool> workers{};
+
+  std::condition_variable queue_or_death;
+  synchronized<std::optional<std::deque<task>>> queue{std::in_place};
+
+  std::size_t tasks_submitted{0};
+  std::condition_variable on_task_completion;
+  synchronized<std::size_t> tasks_completed{0u};
+
+  static void pool_work(std::size_t id, thread_pool* pool);
+  static void borrow_then_pool_work(task&& t,
+                                    std::size_t id,
+                                    thread_pool* pool);
+
+  std::optional<task> dequeue();
+  std::optional<task> finish_and_dequeue();
+};
+
+template <typename F>
+auto enqueue(thread_pool& pool, F&& f) -> std::future<std::invoke_result_t<F>>
+{
+  using result_type = std::invoke_result_t<F>;
+  // todo(ssura): in c++23 we can use std::move_only_function
+  //              and pass the promise directly into the function.
+  //              currently this approach does not work because
+  //              std::function requires the function to be copyable,
+  //              but std::promise obviously is not.
+  std::shared_ptr p = std::make_shared<std::promise<result_type>>();
+  std::future fut = p->get_future();
+  pool.enqueue([f = std::move(f), mp = std::move(p)]() mutable {
+    try {
+      if constexpr (std::is_same_v<result_type, void>) {
+        f();
+        mp->set_value();
+      } else {
+        mp->set_value(f());
+      }
+    } catch (...) {
+      mp->set_exception(std::current_exception());
+    }
+  });
```

Comment:
Its important to note that i package a function f returning a B into a wrapper that returns nothing while still returning a std::future\<B\>.
This is not pretty to do with `std::packaged_task` since you either
* return a std::future\<void\> with std::packaged_task\<void\>
* return a std::future\<B\> with std::packaged_task\<B\>, but you now have to somehow create a type erased wrapper for the task.  std::function will not work since it requires being copyable.

I have not found a way yet to make this with packaged_task while still being somewhat easy to understand and read.
It should be doable with `std::future::then` but thats still in experimental so not even in C++23.